crates/sui-data-ingestion-core/src/reader.rs

```diff
@@ -61,31 +65,81 @@ impl CheckpointReader {
-                    }
-                });
-            for bytes in try_join_all(futures).await? {
-                if bytes.is_none() {
+    fn exceeds_capacity(&self, checkpoint_number: CheckpointSequenceNumber) -> bool {
+        (MAX_CHECKPOINTS_IN_PROGRESS as u64 + self.last_pruned_watermark) <= checkpoint_number
+    }
+
+    async fn remote_fetch_checkpoint(
+        store: &dyn ObjectStore,
+        checkpoint_number: CheckpointSequenceNumber,
+    ) -> Result<CheckpointData> {
+        let path = Path::from(format!("{}.chk", checkpoint_number));
+        let response = store.get(&path).await?;
+        let bytes = response.bytes().await?;
+        Blob::from_bytes::<CheckpointData>(&bytes)
+    }
+
+    fn start_remote_fetcher(&mut self) -> mpsc::Receiver<Result<CheckpointData>> {
+        let batch_size = self.remote_read_batch_size;
+        let start_checkpoint = self.current_checkpoint_number;
+        let (sender, receiver) = mpsc::channel(batch_size);
+        let url = self
+            .remote_store_url
+            .clone()
+            .expect("remote store url must be set");
+        let store = if self.remote_store_options.is_empty() {
+            let retry_config = RetryConfig {
+                max_retries: 0,
+                retry_timeout: Duration::from_secs(10),
+                ..Default::default()
+            };
+            let http_store = object_store::http::HttpBuilder::new()
+                .with_url(url)
+                .with_retry(retry_config)
+                .build()
+                .expect("failed to parse remote store config");
+            Box::new(http_store)
+        } else {
+            parse_url_opts(
+                &Url::parse(&url).expect("failed to parse remote store url"),
+                self.remote_store_options.clone(),
+            )
+            .expect("failed to parse remote store config")
+            .0
+        };
+
+        spawn_monitored_task!(async move {
+            let mut checkpoint_stream = (start_checkpoint..u64::MAX)
+                .map(|checkpoint_number| Self::remote_fetch_checkpoint(&store, checkpoint_number))
+                .pipe(futures::stream::iter)
+                .buffered(batch_size);
+
+            while let Some(checkpoint) = checkpoint_stream.next().await {
+                if sender.send(checkpoint).await.is_err() {
                     break;
                 }
-                let checkpoint = Blob::from_bytes::<CheckpointData>(&bytes.unwrap())?;
-                checkpoints.push(checkpoint);
             }
+        });
+        receiver
+    }
+
+    fn remote_fetch(&mut self) -> Vec<CheckpointData> {
+        let mut checkpoints = vec![];
+        if self.remote_fetcher_receiver.is_none() {
+            self.remote_fetcher_receiver = Some(self.start_remote_fetcher());
         }
-        Ok(checkpoints)
+        while !self.exceeds_capacity(self.current_checkpoint_number + checkpoints.len() as u64) {
```

Comment:
I'm not sure this does exactly as you expect, you may have some configured "limit" but the underlying fetcher disrespects this and will always try to fetch and enqueue `batch_size*2` number of checkpoints. 1 `batch_size` due to the buffered stream and 1 `batch_size` due to the size of the channel. I suppose if this is sufficiently smaller than MAX_CHECKPOINTS_IN_PROGRESS this could be fine but seems a waste

Reply:
that's a totally different check to overall guarantee that system doesn't process more than `MAX_CHECKPOINTS_IN_PROGRESS` tasks.
batch_size << MAX_CHECKPOINTS_IN_PROGRESS. To give specific numbers currently it's 100 vs 10_000

Reply:
to elaborate more, this is required so we don't attempt to read a message from remote fetcher actor that we are not allowed to process at the moment